### DISCUSSION TITLE
Ingredients/units plural form for other languages

### DISCUSSION THREAD

**Discussion by czubamich** [mealie-recipes/mealie#3214] [Feature Request]
### First Check

- [X] I used the GitHub search to find a similar requests and didn't find it.
- [X] Checked the [tasks tagged](https://github.com/mealie-recipes/mealie/issues?q=is%3Aissue+is%3Aopen+label%3Atask+) issues and verified my feature is not covered

### Please provide a concise description of the problem that would be addressed by this feature.

In languages other than English, plural forms for ingredients and units are more complicated.
To improve the user experience in other languages, it is necessary to implement a feature that allows defining plural forms depending on the rules in the language. And I think this feature would be welcome for polish, russian, arabic, lithuanian etc users.

From what I have noted there can be up to 6 different forms (Zero, One, Two, Few, Many, Other), though it highly depends on the language (in english there are 2 (One, Many), but for polish there are 4 (One, Few, Many, Other).

This feature is needed the most for food ingredients, since for units we can use abbreviations. Though this probably also applies there.

### Please provide a concise description of the feature that would resolve your issue.

I think the simplest solution would be to allow users to define rules for their as shown belown language (initially, although this could be just predefined) and, then if it's defined when configuring a food ingredient, the user could choose to use this feature and provide a list of forms for that food ingredient.

For example for english there would be a simple rule for 2 forms (though we don't need this feature in this case)
`(n ==1) ? {singular} : {many}`
but for polish it would be more complex:
`(n==1 ? {singular} : (n%10>=2 && n%10<=4) && (n%100<12 || n%100>14) ? {few} : n!=1 && (n%10>=0 && n%10<=1) || (n%10>=5 && n%10<=9) || (n%100>=12 && n%100<=14) ? {many} : {other})`

Though for simplicity of implementation instead of {singular},{few},{many} and {other} this function would just return the index of the element in the list we define for the integredient/unit.
Also I think this configuration should be independent of the ingredient/unit name, since variations may not include the form the name appears.

Example eggs (jajka) as countable ingredient, butter and milk as measured with different form than ingredient name (masło, mleko):
![image](https://github.com/mealie-recipes/mealie/assets/64833527/ee1894c1-9973-4929-8647-c6d424298372)
![image](https://github.com/mealie-recipes/mealie/assets/64833527/10b542e3-b803-4ce3-9546-ae028dca9f5e)
![image](https://github.com/mealie-recipes/mealie/assets/64833527/4798d9f0-4b17-4212-902d-bb605a3500a7)

### Please consider and list out some caveats or tradeoffs made in your design decision

**One trade-off is that units can be countable and measurable.**
Again for example in polish if we gave potates as count the would be

- 1 ziemniak, 4 ziemniaki, 8 ziemniaków (1 potato, 4 potatoes, 8 potatoes)

but as measure in grams, teaspoons etc it would be a special form

- 1g ziemniaków, 4g ziemniaków, 8g ziemiaków (1g (of) potatoes, 4g (of) potatoes, 8g (of) potatoes)
- 1g cukru, 4g cukru, 8g cukru (1g (of) sugar, 4g (of) sugar, 8g (of) sugar)
- 1 _szklanka_ cukru, 4 _szklanki_ cukru, 8 _szklanek_ cukru (1 _glass_ (of) sugar, 4 _glasses_ (of) sugar, 8 _glasses_ (of) sugar)

If it was possible to flag the unit behavior as countable or measurable we could only use plural form if not some dedicated form (the one currently implemented), and solve this, though it's up to discussion how it would work.

Also I didn't account for negative amounts, but I guess it's not an issue.

### Additional Information

- [x] If this is accepted I'm willing to submit a PR to provide this feature
- [x] If this is accepted I'm willing to help maintain this feature
- [ ] I'm willing to sponsor/pay a developer to do this work

**Comment by alufers** [comment]
For what it's worth here is a page compiling all of the rules that different languages use:

https://www.unicode.org/cldr/charts/46/supplemental/language_plural_rules.html

**Comment by dbrgn** [comment]
There's an already established pluralization format: ICU MessageFormat http://messageformat.github.io/messageformat/guide/

**Comment by jvalskis** [comment]
Was about to raise the same issue myself. Crowdin seems to support the feature already: https://support.crowdin.com/icu-message-syntax/#plural
I guess it's just a matter of updating relevant resources and invoking translation with proper context
